stdlib/public/Concurrency/Task.cpp

```diff
@@ -21,20 +21,129 @@
 #include "TaskPrivate.h"
 
 using namespace swift;
+using FutureFragment = AsyncTask::FutureFragment;
+
+size_t FutureFragment::fragmentSize(const Metadata *resultType) {
+  return storageOffset(resultType) + resultType->vw_size();
+}
+
+void FutureFragment::destroy() {
+  switch (status.load()) {
+  case Status::Executing:
+    assert(false && "destroying a task that never completed");
+
+  case Status::Success:
+    resultType->vw_destroy(getStoragePtr());
+    break;
+
+  case Status::Error:
+    swift_unknownObjectRelease(getStoragePtr());
+    break;
+  }
+}
+
+FutureFragment::Status AsyncTask::waitFuture(AsyncTask *waitingTask) {
+  assert(isFuture());
+  auto fragment = futureFragment();
+
+  auto currentStatus = fragment->status.load();
+  switch (currentStatus) {
+  case FutureFragment::Status::Error:
+  case FutureFragment::Status::Success:
+    // The task is done; we don't need to wait.
+    return currentStatus;
+
+  case FutureFragment::Status::Executing:
+    break;
+  }
+
+  // Put the waiting task at the beginning of the wait queue.
+  while (true) {
+    waitingTask->NextWaitingTask = fragment->waitQueue.load();
+    if (fragment->waitQueue.compare_exchange_strong(
+            waitingTask->NextWaitingTask, waitingTask)) {
```

Comment:
The reason you should unify the two atomic fields is that you have a race here against someone concurrently completing the task.  The thread that completes the task should exchange the appropriate completion value into the queue and then process the waiting tasks they just loaded (via the exchange).  Here you need to loop trying to add the task to the queue until you see success.

Always use memory orderings when working with `std::atomic`.  This needs to be a `std::memory_order_release` on success; it can be `std::memory_order_relaxed` on failure, except actually it needs to be `std::memory_order_acquire` because you're going to check for success/error.

Don't repeatedly reload `waitQueue`.  Load it (with an acquire) into a local, then use that local in the `compare_exchange`.  When you see a pointer, copy it into `NextWaitingTask` (or really the `SchedulerPrivate` contents).

You should use `compare_exchange_weak` since you're already doing this in a loop.

Reply:
Thank you! I've made all of these changes together. It looks much more reasonable now.